DESIGN.md

```diff
@@ -0,0 +1,387 @@
+)
+
+var (
+    AlgorithmPS256   *Algorithm
+    AlgorithmPS384   *Algorithm
+    AlgorithmPS512   *Algorithm
+    AlgorithmES256   *Algorithm
+    AlgorithmES384   *Algorithm
+    AlgorithmES512   *Algorithm
+    AlgorithmEd25519 *Algorithm
+)
+
+type Algorithm struct {
+	Name     string
+	Value    int
+	HashFunc crypto.Hash
+}
+
+func I2OSP(x *big.Int, n int) ([]byte, error)
+func OS2IP(x []byte) *big.Int
+
+type ProtectedHeader map[interface{}]interface{}
+    func (h *ProtectedHeader) MarshalCBOR() ([]byte, error)
+    func (h *ProtectedHeader) UnmarshalCBOR(data []byte) error
+    func (h *ProtectedHeader) SetAlgorithm(alg *Algorithm)
+
+type UnprotectedHeader map[interface{}]interface{}
+
+type Headers struct {
+    RawProtected   []byte
+    RawUnprotected []byte
+	Protected      ProtectedHeader
+	Unprotected    UnprotectedHeader
+}
+
+type Signature struct {
+	Headers   Headers
+	Signature []byte
+}
+    func NewSignature() *Signature
+    func (s *Signature) MarshalCBOR() ([]byte, error)
+    func (s *Signature) UnmarshalCBOR(data []byte) error
+
+type SignMessage struct {
+    Headers    Headers
+    Payload    []byte
+    Signatures []*Signature
+}
+    func NewSignMessage() *SignMessage
+    func ParseSignMessage(data []byte) (*SignMessage, error)
+    func (m *SignMessage) MarshalCBOR() ([]byte, error)
+    func (m *SignMessage) UnmarshalCBOR(data []byte) error
+    func (m *SignMessage) Sign(ctx context.Context, external []byte, signers []Signer) error
+    func (m *SignMessage) Verify(ctx context.Context, external []byte, verifier []Verifier) error
+
+type Sign1Message struct {
+    Headers   Headers
+    Payload   []byte
+    Signature []byte
+}
+    func NewSign1Message() *Sign1Message
+    func ParseSign1Message(data []byte) (*Sign1Message, error)
+    func (m *Sign1Message) MarshalCBOR() ([]byte, error)
+    func (m *Sign1Message) UnmarshalCBOR(data []byte) error
+    func (m *Sign1Message) Sign(ctx context.Context, external []byte, signer Signer) error
+    func (m *Sign1Message) Verify(ctx context.Context, external []byte, verifier Verifier) error
+
+type Signer interface {
+    Algorithm() *Algorithm
+    Sign(ctx context.Context, digest []byte) ([]byte, error)
+}
+    func NewSigner(alg *Algorithm, key crypto.PrivateKey) (Signer, error)
+    func NewSignerWithEphemeralKey(alg *Algorithm) (Signer, crypto.PrivateKey, error)
+
+type Verifier interface {
+    Algorithm() *Algorithm
+	Verify(ctx context.Context, digest, signature []byte) error
+}
+    func NewVerifier(alg *Algorithm, key crypto.PublicKey) (Verifier, error)
+```
+
+Key changes:
+
+- Header related functions are cleaned up.
+  - String header labels are not registered in [IANA "COSE Header Parameters" Registry](https://www.iana.org/assignments/cose/cose.xhtml#header-parameters), and thus are cleaned.
+- `Signer` is now an interface.
+  - `Sign` takes `context.Context` as input with `rand io.Reader` removed since it is potentially a RPC call.

```

Comment:
Let's add `rand` back as some devices do not have stable entropy source.

Most users should use platform [CSPRNGs](https://en.wikipedia.org/wiki/Cryptographically-secure_pseudorandom_number_generator) provided by [crypto/rand](https://pkg.go.dev/crypto/rand). Crypto experts may opt to deterministic signatures (e.g. [RFC6979](https://datatracker.ietf.org/doc/html/rfc6979)).

Reply:
The [godoc](https://pkg.go.dev/context) of the `context` package states that

> Do not store Contexts inside a struct type; instead, pass a Context explicitly to each function that needs it.

It is suggested to have `ctx` per API call.

Reply:
Combining the above two comments, we will have something like 

```go
type Signer interface {
    Algorithm() *Algorithm
    Sign(ctx context.Context, rand io.Reader, digest []byte) ([]byte, error)
}

func (m *Sign1Message) Sign(ctx context.Context, rand io.Reader, external []byte, signer Signer) error
```

and we have an UX issue here. Usually, local signing does not require context and remote signing does not require random source.

Reply:
In the end, we might end up with

```golang
type Signer interface {
    Algorithm() *Algorithm
    Sign(rand io.Reader, digest []byte) ([]byte, error)
}

type ContextSigner interface {
    Algorithm() *Algorithm
    SignWithContext(ctx context.Context, digest []byte) ([]byte, error)
}

func (m *Sign1Message) Sign(ctx context.Context, rand io.Reader, external []byte, signer Signer) error
func (m *Sign1Message) SignWithContext(ctx context.Context, external []byte, signer ContextSigner) error
```

Reply:
@thomas-fossati Do you have any thoughts?

Reply:
> The [godoc](https://pkg.go.dev/context) of the `context` package states that
> 
> > Do not store Contexts inside a struct type; instead, pass a Context explicitly to each function that needs it.
> 
> It is suggested to have `ctx` per API call.

Sure, but a fresh new context could be passed to the signer when the function is called, e.g. using a `WithContext` method:
```
 ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
 defer cancel()

 sig, err := jws.Sign(tbs, algo, signer.WithContext(ctx))
```

Reply:
[Request.WithContext(ctx context.Context)](https://pkg.go.dev/net/http#Request.WithContext) in `http` package is also one approach.

Reply:
However, it would be better if `context.Context` can be promoted to the first class since the remote signing scenarios would be common in CI/CD pipelines as private keys are typically should not be stored on the local machine but in a KMS or HSM.

Reply:
> Let's add `rand` back as some devices do not have stable entropy source.
> 
> Most users should use platform [CSPRNGs](https://en.wikipedia.org/wiki/Cryptographically-secure_pseudorandom_number_generator) provided by [crypto/rand](https://pkg.go.dev/crypto/rand). Crypto experts may opt to deterministic signatures (e.g. [RFC6979](https://datatracker.ietf.org/doc/html/rfc6979)).

I thought about deterministic signatures as in the previous comment. After reading through the golang implementation, I found that the implementation of `crypto/ecdsa` does not allow deterministic ECDSA as it [randomizes the random source](https://github.com/golang/go/blob/go1.18/src/crypto/ecdsa/ecdsa.go#L195). For the [FIPS-compliant version](https://github.com/golang/go/blob/dev.boringcrypto.go1.18/src/crypto/ecdsa/ecdsa.go#L224) of golang, BoringSSL does not allow any other random sources except the one in `rand.Reader`. I don't see another values other than `rand.Reader` in `crypto.rand` can be passed in as `rand`. 

@thomas-fossati Could you provide one scenario that we should preserve `rand`?

Reply:
A "0"-reader for deterministic output in testing is not suggested and it can really make dead loops. I also reviewed the top `rand` packages searched from `pkg.go.dev`. Except `crypto/rand`, none of them are CSPNRGs, and should not be used for crypto.

Nonetheless, I do agree with you on providing  _crypto agility_ to some extent due to bad or buggy platform default implementations. Additionally, I found a conversation at https://github.com/golang/go/issues/38548 about the golang crypto behavior. Filippo Valsorda, who is the owner of the golang built-in packages `crypto/...`, commented that

> I don't disagree that it would be even better if we did not accept a Reader for the randomness, but unfortunately some applications have opinions on their CSPRNGs.

Since `go-cose` will be used in low-level as the `crypto` packages, we do need to align that crypto agility.

Reply:
Updated with 815420c. /cc @thomas-fossati